File: src-cpp/leetcode/letter-combinations-of-a-phone-number.cpp
#include <bits/stdc++.h>
using namespace std;

const vector<string> telepad = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

// iterative
// building on top of prev result
vector<string> letterCombinations(string digits)
{
    if (digits.empty())
        return {};

    vector<string> res;

    for (auto digit : digits)
    {
        vector<string> nextRes;
        for (auto candidate : telepad[digit - '2'])
        {
            // build next combinations using prev result
            for (auto str : res)
                nextRes.push_back(str + candidate);
        }

        swap(res, nextRes);
    }

    return res;
}

// recursive
vector<string> letterCombinations(string digits)
{
    if (digits.empty())
        return {};

    vector<string> res;
    const vector<string> telepad = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    function<void(string &, int, string)> helper =
        [&res, &helper, &telepad](string &digits, int i, string combi)
    {
        if (i == size(digits))
        {
            res.push_back(move(combi));
            return;
        }

        for (auto candidate : telepad[digits[i] - '2'])
            helper(digits, i + 1, combi + candidate);
    };

    helper(digits, 0, "");
    return res;
}

File: src-ts/leetcode/letter-combinations-of-a-phone-number.ts
const telepad = ['abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz'];

// iterative
function letterCombinations(digits: string): string[] {
  if (digits.length === 0) return [];

  let res: string[] = [''];

  for (const digit of digits) {
    const nextRes: string[] = [];

    for (const char of telepad[Number(digit) - 2]) {
      // build next combinations using prev result
      for (const str of res) {
        nextRes.push(str + char);
      }
    }

    res = nextRes;
  }

  return res;
}

// recursive
function letterCombinations(digits: string): string[] {
  if (digits.length === 0) return [];

  const res: string[] = [];

  const helper = (index: number, combination: string) => {
    if (index === digits.length) {
      res.push(combination);
      return;
    }

    for (const char of telepad[Number(digits[index]) - 2]) {
      helper(index + 1, combination + char);
    }
  };

  helper(0, '');

  return res;
}
